network/framework/src/protocols/wire/messaging/v1/metadata.rs

```diff
@@ -0,0 +1,234 @@
+    StreamedMessageFragment, // A fragment of a streamed message (not the head or tail)
+    StreamedMessageTail,     // The tail (last fragment) of a streamed message
+}
+
+impl MessageStreamType {
+    pub fn get_label(&self) -> &'static str {
+        match self {
+            MessageStreamType::NonStreamedMessage => "NonStreamedMessage",
+            MessageStreamType::StreamedMessageHead => "StreamedMessageHead",
+            MessageStreamType::StreamedMessageFragment => "StreamedMessageFragment",
+            MessageStreamType::StreamedMessageTail => "StreamedMessageTail",
+        }
+    }
+}
+
+/// A struct holding metadata about each message.
+/// Note: this is not sent along the wire, it is only used internally.
+#[derive(Clone, Debug, PartialEq, Eq, Deserialize, Serialize)]
+pub struct MessageMetadata {
+    /// The network ID for the message
+    network_id: NetworkId,
+
+    /// The protocol ID for the message. This may not always be
+    /// known (e.g., when failing to deserialize a message).
+    protocol_id: Option<ProtocolId>,
+
+    /// The type of message being sent
+    message_send_type: MessageSendType,
+
+    /// The stream type of the message being sent
+    message_stream_type: MessageStreamType,
+
+    /// The time at which the message was sent by the application
+    application_send_time: Option<SystemTime>,
+}
+
+impl MessageMetadata {
+    pub fn new(
+        network_id: NetworkId,
+        protocol_id: Option<ProtocolId>,
+        message_send_type: MessageSendType,
+        application_send_time: Option<SystemTime>,
+    ) -> Self {
+        Self {
+            network_id,
+            protocol_id,
+            message_send_type,
+            message_stream_type: MessageStreamType::NonStreamedMessage, // Default to non-streamed messages
+            application_send_time,
+        }
+    }
+
+    /// Returns the time at which the message was first sent by the application
+    pub fn application_send_time(&self) -> Option<SystemTime> {
+        self.application_send_time
+    }
+
+    /// Marks the message as being sent over the network wire by emitting latency metrics
+    pub fn mark_message_as_sending(&mut self) {
+        // If this message is a streamed message fragment, there's no need to emit
+        // any metrics (we only emit metrics for the head and tail of streamed messages).
+        if self.message_stream_type == MessageStreamType::StreamedMessageFragment {
+            return;
+        }
+
+        // Otherwise, emit the latency metrics
+        if let Some(application_send_time) = self.application_send_time {
+            // Calculate the application to wire send latency
+            let application_to_wire_latency = application_send_time
+                .elapsed()
+                .unwrap_or_default()
+                .as_secs_f64();
+
+            // Update the application to wire latency metrics
+            counters::observe_message_send_latency(
```

Comment:
this is the pre-send latency (more like queueing time) right? it might be worth recording the post-send latency too (after https://github.com/aptos-labs/aptos-core/blob/1099aec2270c6f5865bb2f6bcbe1cf483c18a7d5/network/framework/src/peer/mod.rs#L361)

Reply:
Gotcha! Added another metric for the post-send latency. So now, before the send, we save the "wire start send time", and then once the message finishes being "sent", we log the "wire send time", too. Let me know if this is what you meant 😄